server/unithand.cpp

```diff
@@ -4022,146 +4020,213 @@ static bool do_attack(struct unit *punit, struct tile *def_tile,
+                  defender_vet, defender_link, pdefender->id, defender_fp,
                   (float) def_power / POWER_FACTOR, def_hp_start,
                   nation_adjective_for_player(unit_owner(punit)),
-                  attacker_vet, winner_link, punit->id, attacker_fp,
-                  (float) att_power / POWER_FACTOR,
-                  att_hp_start - pwinner->hp, pwinner->hp,
-                  vet ? unit_achieved_rank_string(punit) : "");
-
-    notify_player(
-        unit_owner(punit), unit_tile(punit), E_UNIT_WIN_ATT, ftc_server,
-        /* TRANS: "Your attacking green Legion [id:200 ...A:4.0
-         * lost 1 HP, has 9 HP remaining] succeeded against the
-         * Greek green Warriors [id:100 HP:10]." */
-        _("Your attacking %s %s [id:%d %s%sA:%.1f lost %d HP, "
-          "has %d remaining] succeeded against the %s %s %s "
-          "[id:%d HP:%d]."),
-        attacker_vet, winner_link, punit->id, attacker_fp, attacker_tired,
-        (float) att_power / POWER_FACTOR, att_hp_start - pwinner->hp,
-        pwinner->hp, nation_adjective_for_player(unit_owner(pdefender)),
-        defender_vet, loser_link, pdefender->id, def_hp_start);
+                  attacker_vet, attacker_link, punit->id, attacker_fp,
+                  attacker_tired, (float) att_power / POWER_FACTOR,
+                  att_hp_start - punit->hp, punit->hp,
+                  (punit->veteran - old_unit_vet)
+                      ? unit_achieved_rank_string(punit)
+                      : "");
 
     punit->moved = true; // We moved
-    kill_unit(pwinner, ploser,
-              vet && !utype_is_consumed_by_action(paction, punit->utype));
+    kill_unit(punit, pdefender,
+              punit->veteran == old_unit_vet
+                  && !utype_is_consumed_by_action(paction, punit->utype));
 
     /* Now that dead defender is certainly no longer listed as unit
      * supported by the city, we may even remove the city
      * (if it shrinks from size 1) */
     if (auto pcity = tile_city(def_tile); pcity != nullptr) {
       unit_attack_civilian_casualties(punit, pcity, paction, "attack");
     }
-    if (unit_is_alive(winner_id)) {
-      if (utype_is_consumed_by_action(paction, pwinner->utype)) {
-        return true;
+    if (!utype_is_consumed_by_action(paction, punit->utype)) {
+      /* If attacker wins, and occupychance > 0, it might move in.  Don't
+       * move in if there are enemy units in the tile (a fortress, city or
+       * air base with multiple defenders and unstacked combat). Note that
+       * this could mean capturing (or destroying) a city. */
+      if (fc_rand(100) < game.server.occupychance
+          && !is_non_allied_unit_tile(def_tile, pplayer)) {
+        occupy_move(punit, def_tile);
+      }
+      // The attacker may have died for many reasons
+      // 2024-09-04 HF: Not very sure why this is needed here
+      // but not in the other outcome cases? Is send_combat()
+      // doing this as well?
```

Comment:
For instance if there is a hut on the target tile.

Reply:
updated comment